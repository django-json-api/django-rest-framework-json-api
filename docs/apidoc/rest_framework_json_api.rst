rest\_framework\_json\_api package
==================================

.. automodule:: rest_framework_json_api
    :members:
    :undoc-members:
    :show-inheritance:

Subpackages
-----------

.. toctree::

    rest_framework_json_api.django_filters
    rest_framework_json_api.filters

Submodules
----------

.. toctree::

   rest_framework_json_api.exceptions
   rest_framework_json_api.metadata
   rest_framework_json_api.mixins
   rest_framework_json_api.pagination
   rest_framework_json_api.parsers
   rest_framework_json_api.relations
   rest_framework_json_api.renderers
   rest_framework_json_api.serializers
   rest_framework_json_api.settings
   rest_framework_json_api.utils
   rest_framework_json_api.views

