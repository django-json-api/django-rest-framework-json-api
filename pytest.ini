[pytest]
DJANGO_SETTINGS_MODULE=example.settings
