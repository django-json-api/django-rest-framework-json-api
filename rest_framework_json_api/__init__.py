# -*- coding: utf-8 -*-

__title__ = 'djangorestframework-jsonapi'
__version__ = '2.3.0'
__author__ = ''
__license__ = 'MIT'
__copyright__ = ''

# Version synonym
VERSION = __version__
