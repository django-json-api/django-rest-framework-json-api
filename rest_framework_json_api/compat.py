# Django REST framework 3.14 removed NullBooleanField
# can be removed once support for DRF 3.13 is dropped.
try:
    from rest_framework.serializers import NullBooleanField
except ImportError:  # pragma: no cover
    NullBooleanField = object()


# Django REST framework 3.14 deprecates usage of `_get_reference`.
# can be removed once support for DRF 3.13 is dropped.
def get_reference(schema, serializer):
    try:
        return schema.get_reference(serializer)
    except AttributeError:  # pragma: no cover
        return schema._get_reference(serializer)
