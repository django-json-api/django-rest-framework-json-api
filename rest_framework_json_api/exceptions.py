from django.utils.translation import ugettext_lazy as _
from rest_framework import exceptions, status

from rest_framework_json_api import utils

from .settings import json_api_settings


def rendered_with_json_api(view):
    from rest_framework_json_api.renderers import JSONRenderer
    for renderer_class in getattr(view, 'renderer_classes', []):
        if issubclass(renderer_class, JSONRenderer):
            return True
    return False


def exception_handler(exc, context):
    # Import this here to avoid potential edge-case circular imports, which
    # crashes with:
    # "ImportError: Could not import 'rest_framework_json_api.parsers.JSONParser' for API setting
    # 'DEFAULT_PARSER_CLASSES'. ImportError: cannot import name 'exceptions'.'"
    #
    # Also see: https://github.com/django-json-api/django-rest-framework-json-api/issues/158
    from rest_framework.views import exception_handler as drf_exception_handler

    # Render exception with DRF
    response = drf_exception_handler(exc, context)
    if not response:
        return response

    # Use regular DRF format if not rendered by DRF JSON API and not uniform
    is_json_api_view = rendered_with_json_api(context['view'])
    is_uniform = json_api_settings.UNIFORM_EXCEPTIONS
    if not is_json_api_view and not is_uniform:
        return response

    # Convert to DRF JSON API error format
    response = utils.format_drf_errors(response, context, exc)

    # Add top-level 'errors' object when not rendered by DRF JSON API
    if not is_json_api_view:
        response.data = utils.format_errors(response.data)

    return response


class Conflict(exceptions.APIException):
    status_code = status.HTTP_409_CONFLICT
    default_detail = _('Conflict.')
