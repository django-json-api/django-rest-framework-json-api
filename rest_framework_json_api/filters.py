import re

from rest_framework.exceptions import ValidationError
from rest_framework.filters import BaseFilterBackend, OrderingFilter

from rest_framework_json_api.utils import format_value


class OrderingFilter(OrderingFilter):
    """
    A backend filter that implements http://jsonapi.org/format/#fetching-sorting and
    raises a 400 error if any sort field is invalid.

    If you prefer *not* to report 400 errors for invalid sort fields, just use
    :py:class:`rest_framework.filters.OrderingFilter` with
    :py:attr:`~rest_framework.filters.OrderingFilter.ordering_param` = "sort"

    Also applies DJA format_value() to convert (e.g. camelcase) to underscore.
    (See JSON_API_FORMAT_FIELD_NAMES in docs/usage.md)
    """
    #: override :py:attr:`rest_framework.filters.OrderingFilter.ordering_param`
    #: with JSON:API-compliant query parameter name.
    ordering_param = 'sort'

    def remove_invalid_fields(self, queryset, fields, view, request):
        """
        Extend :py:meth:`rest_framework.filters.OrderingFilter.remove_invalid_fields` to
        validate that all provided sort fields exist (as contrasted with the super's behavior
        which is to silently remove invalid fields).

        :raises ValidationError: if a sort field is invalid.
        """
        valid_fields = [
            item[0] for item in self.get_valid_fields(queryset, view,
                                                      {'request': request})
        ]
        bad_terms = [
            term for term in fields
            if format_value(term.replace(".", "__").lstrip('-'), "underscore") not in valid_fields
        ]
        if bad_terms:
            raise ValidationError('invalid sort parameter{}: {}'.format(
                ('s' if len(bad_terms) > 1 else ''), ','.join(bad_terms)))
        # this looks like it duplicates code above, but we want the ValidationError to report
        # the actual parameter supplied while we want the fields passed to the super() to
        # be correctly rewritten.
        # The leading `-` has to be stripped to prevent format_value from turning it into `_`.
        underscore_fields = []
        for item in fields:
            item_rewritten = item.replace(".", "__")
            if item_rewritten.startswith('-'):
                underscore_fields.append(
                    '-' + format_value(item_rewritten.lstrip('-'), "underscore"))
            else:
                underscore_fields.append(format_value(item_rewritten, "underscore"))

        return super(OrderingFilter, self).remove_invalid_fields(
            queryset, underscore_fields, view, request)


class QueryParameterValidationFilter(BaseFilterBackend):
    """
    A backend filter that performs strict validation of query parameters for
    JSON:API spec conformance and raises a 400 error if non-conforming usage is
    found.

    If you want to add some additional non-standard query parameters,
    override :py:attr:`query_regex` adding the new parameters. Make sure to comply with
    the rules at http://jsonapi.org/format/#query-parameters.
    """
    #: compiled regex that matches the allowed http://jsonapi.org/format/#query-parameters:
    #: `sort` and `include` stand alone; `filter`, `fields`, and `page` have []'s
    query_regex = re.compile(r'^(sort|include)$|^(filter|fields|page)(\[[\w\.\-]+\])?$')

    def validate_query_params(self, request):
        """
        Validate that query params are in the list of valid query keywords in
        :py:attr:`query_regex`

        :raises ValidationError: if not.
        """
        # TODO: For jsonapi error object conformance, must set jsonapi errors "parameter" for
        # the ValidationError. This requires extending DRF/DJA Exceptions.
        for qp in request.query_params.keys():
            if not self.query_regex.match(qp):
                raise ValidationError('invalid query parameter: {}'.format(qp))
            if len(request.query_params.getlist(qp)) > 1:
                raise ValidationError(
                    'repeated query parameter not allowed: {}'.format(qp))

    def filter_queryset(self, request, queryset, view):
        """
        Overrides :py:meth:`BaseFilterBackend.filter_queryset` by first validating the
        query params with :py:meth:`validate_query_params`
        """
        self.validate_query_params(request)
        return queryset
