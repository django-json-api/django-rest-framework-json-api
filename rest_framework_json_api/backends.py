from rest_framework.exceptions import ValidationError
from rest_framework.filters import OrderingFilter

from rest_framework_json_api.utils import format_value


class JSONAPIOrderingFilter(OrderingFilter):
    """
    This implements http://jsonapi.org/format/#fetching-sorting and raises 400
    if any sort field is invalid. If you prefer *not* to report 400 errors for
    invalid sort fields, just use OrderingFilter with `ordering_param='sort'`

    TODO: Add sorting based upon relationships (sort=relname.fieldname)
    """
    ordering_param = 'sort'

    def remove_invalid_fields(self, queryset, fields, view, request):
        """
        overrides remove_invalid_fields to raise a 400 exception instead of
        silently removing them. set `ignore_bad_sort_fields = True` to not
        do this validation.
        """
        valid_fields = [
            item[0] for item in self.get_valid_fields(queryset, view,
                                                      {'request': request})
        ]
        bad_terms = [
            term for term in fields
            if format_value(term.lstrip('-'), "underscore") not in valid_fields
        ]
        if bad_terms:
            raise ValidationError('invalid sort parameter{}: {}'.format(
                ('s' if len(bad_terms) > 1 else ''), ','.join(bad_terms)))

        return super(JSONAPIOrderingFilter, self).remove_invalid_fields(
            queryset, fields, view, request)
