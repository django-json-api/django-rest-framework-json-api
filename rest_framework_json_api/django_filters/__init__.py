from .backends import DjangoFilterBackend  # noqa: F401
