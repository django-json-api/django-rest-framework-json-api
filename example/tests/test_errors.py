import pytest
from django.test import override_settings
from django.urls import path, reverse
from rest_framework import generics

from rest_framework_json_api import serializers

from example.models import Author, Blog


# serializers
class CommentAttachmentSerializer(serializers.Serializer):
    data = serializers.CharField(allow_null=False, required=True)

    def validate_data(self, value):
        if value and len(value) < 10:
            raise serializers.ValidationError("Too short data")


class CommentSerializer(serializers.Serializer):
    attachments = CommentAttachmentSerializer(many=True, required=False)
    attachment = CommentAttachmentSerializer(required=False)
    one_more_attachment = CommentAttachmentSerializer(required=False)
    body = serializers.CharField(allow_null=False, required=True)


class EntrySerializer(serializers.Serializer):
    blog = serializers.IntegerField()
    comments = CommentSerializer(many=True, required=False)
    comment = CommentSerializer(required=False)
    headline = serializers.CharField(allow_null=True, required=True)
    body_text = serializers.CharField()
    author = serializers.ResourceRelatedField(
        queryset=Author.objects.all(), required=False
    )

    def validate(self, attrs):
        body_text = attrs["body_text"]
        if len(body_text) < 5:
            raise serializers.ValidationError(
                {"body_text": {"title": "Too Short title", "detail": "Too short"}}
            )


# view
class DummyTestView(generics.CreateAPIView):
    serializer_class = EntrySerializer
    resource_name = "entries"

    def get_serializer_context(self):
        return {}


urlpatterns = [
    path("entries-nested", DummyTestView.as_view(), name="entries-nested-list")
]


@pytest.fixture(scope="function")
def some_blog(db):
    return Blog.objects.create(name="Some Blog", tagline="It's a blog")


def perform_error_test(client, data):
    with override_settings(ROOT_URLCONF=__name__):
        url = reverse("entries-nested-list")
        response = client.post(url, data=data)

    return response.json()


def test_first_level_attribute_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
            },
        }
    }
    assert snapshot == perform_error_test(client, data)


def test_first_level_custom_attribute_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "body-text": "body",
                "headline": "headline",
            },
        }
    }
    with override_settings(JSON_API_FORMAT_FIELD_NAMES="dasherize"):
        assert snapshot == perform_error_test(client, data)


def test_second_level_array_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comments": [{}],
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_second_level_dict_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comment": {},
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_third_level_array_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comments": [{"body": "test comment", "attachments": [{}]}],
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_third_level_custom_array_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comments": [
                    {"body": "test comment", "attachments": [{"data": "text"}]}
                ],
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_third_level_dict_error(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comments": [{"body": "test comment", "attachment": {}}],
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_many_third_level_dict_errors(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
                "comments": [{"attachment": {}}],
            },
        }
    }

    assert snapshot == perform_error_test(client, data)


def test_relationship_errors_has_correct_pointers(client, some_blog, snapshot):
    data = {
        "data": {
            "type": "entries",
            "attributes": {
                "blog": some_blog.pk,
                "bodyText": "body_text",
                "headline": "headline",
            },
            "relationships": {
                "author": {"data": {"id": "INVALID_ID", "type": "authors"}}
            },
        }
    }

    assert snapshot == perform_error_test(client, data)
