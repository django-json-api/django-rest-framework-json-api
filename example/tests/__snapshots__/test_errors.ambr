# name: test_first_level_attribute_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/headline',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_first_level_custom_attribute_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'detail': 'Too short',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/body-text',
        },
        'title': 'Too Short title',
      },
    ],
  }
---
# name: test_many_third_level_dict_errors
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/attachment/data',
        },
        'status': '400',
      },
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/body',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_relationship_errors_has_correct_pointers
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'incorrect_type',
        'detail': 'Incorrect type. Expected resource identifier object, received str.',
        'source': <class 'dict'> {
          'pointer': '/data/relationships/author',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_second_level_array_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/body',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_second_level_dict_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comment/body',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_third_level_array_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/attachments/0/data',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_third_level_custom_array_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'invalid',
        'detail': 'Too short data',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/attachments/0/data',
        },
        'status': '400',
      },
    ],
  }
---
# name: test_third_level_dict_error
  <class 'dict'> {
    'errors': <class 'list'> [
      <class 'dict'> {
        'code': 'required',
        'detail': 'This field is required.',
        'source': <class 'dict'> {
          'pointer': '/data/attributes/comments/0/attachment/data',
        },
        'status': '400',
      },
    ],
  }
---
